### PR TITLE
Speed up full scan completion in IRB

### PR DESCRIPTION
**Pull request by dgutov**
This greatly improves completion performance in the full scan case (when IRB doesn't know the type of the call target).

It's particularly noticeable in large projects. Here's how to emulate one:

```rb
N = 40000

# Save the references to avoid GC.
modules = (1..N).map do |i|
  Module.new do
    module_eval <<-DEF
      def foo#{i}
      end
DEF
  end
end; nil
```

In such an environment, this branch brings completion time from ~1 second to almost instantaneous. This code (or close enough) has already been added to Pry and released in version 0.11.0. Here's the relevant report there: https://github.com/pry/pry/issues/1540

### CODE REVIEW

**Comment by dgutov** [comment]
@yuki24 Any suggestions for how to invoke completion programmatically?

**Comment by yuki24** [comment]
I'm not familiar with IRB but I think you may be able to use the `IRB::InputCompletor::CompletionProc` directly.

**Comment by dgutov** [comment]
@yuki24 Yup, it was easier than I expected. I'll take out a commit I'm unsure about now and post the resulting comparison.

**Comment by dgutov** [comment]
Here's the benchmark:

```rb
require 'irb/completion'
require 'benchmark'

N = 40000

# Save the references to avoid GC.
modules = (1..N).map do |i|
  Module.new do
    module_eval <<-DEF
      def foo#{i}
      end
DEF
  end
end; nil

puts Benchmark.measure {
  IRB::InputCompletor::CompletionProc.call('foo.bar.')
}
```

And here are the results:

|      |N 20000 |N 40000 |
|------|--------|--------|
|trunk |  0.586|  1.105 |
|this patch | 0.079  | 0.138  |

The difference is smaller than with older rubies I've tested this before, but still very significant.

**Comment by dgutov** [comment]
@yuki24 Is this something you can review?

**Comment by yuki24** [comment]
The benchmark looks feasible, but I don't have a final say to this. Let's see how other maintainers think of this.

File: lib/irb/completion.rb [hunk 1]
```diff
@@ -218,6 +209,33 @@ def self.select_message(receiver, message, candidates, sep = ".")
...
+      scanner = lambda do |m|
+        next if s.include?(m) # IRB::ExtendCommandBundle::EXCB recurses.
+        s << m
+        m.constants(false).each do |c|
+          value = m.const_get(c)
+          scanner.call(value) if value.is_a?(Module)
+        end
+      end
+
+      %w(IRB SLex RubyLex RubyToken).each do |module_name|
```
Comment (Maumagnaguagno) [inline lib/irb/completion.rb]: You can use ``%i(IRB SLex RubyLex RubyToken)`` to create an array of symbols.

Comment (dgutov) [reply lib/irb/completion.rb]: Good idea, changed.

**Comment by yuki24** [comment]
@Maumagnaguagno Please squash your commits into a single commit when appropriate. This simplifies future cherry picks and keeps the git log clean.

**Comment by Maumagnaguagno** [comment]
@yuki24 I believe your last comment is meant for @dgutov, I have no commits in this pull request.

**Comment by yuki24** [comment]
@Maumagnaguagno My apologies. Yes, that was meant to @dgutov.

**Comment by yuki24** [comment]
I'll merge this PR once the commits are squash into a single commit.

**Comment by dgutov** [comment]
@yuki24 Squashed and force-pushed.

**Pull request closed without merging**